src/Control/Effect/Writer.hs

```diff
@@ -37,93 +42,71 @@ instance Effect (Writer w) where
-runWriter m = runWriterC (interpret m) mempty
+runWriter :: Monoid w => WriterC w m a -> m (w, a)
+runWriter = runState mempty . runWriterC
 {-# INLINE runWriter #-}
 
 -- | Run a 'Writer' effect with a 'Monoid'al log, producing the final log and discarding the result value.
 --
 --   prop> run (execWriter (tell (Sum a) *> pure b)) == Sum a
-execWriter :: forall w sig m a . (Carrier sig m, Effect sig, Monad m, Monoid w) => Eff (WriterC w m) a -> m w
+execWriter :: (Monoid w, Functor m) => WriterC w m a -> m w
 execWriter = fmap fst . runWriter
 {-# INLINE execWriter #-}
 
 
 -- | A space-efficient carrier for 'Writer' effects.
 --
 --   This is based on a post Gabriel Gonzalez made to the Haskell mailing list: https://mail.haskell.org/pipermail/libraries/2013-March/019528.html
---
---   Note that currently, the constant-space behaviour observed there only occurs when using 'WriterC' and 'VoidC' without 'Eff' wrapping them. See the @benchmark@ component for details.
-newtype WriterC w m a = WriterC { runWriterC :: w -> m (w, a) }
-
-instance Functor m => Functor (WriterC w m) where
-  fmap f (WriterC run) = WriterC (\ w -> fmap (fmap f) (run w))
-  {-# INLINE fmap #-}
-
-instance (Monad m, Monoid w) => Applicative (WriterC w m) where
-  pure a = WriterC $ \w -> pure (w, a)
-  {-# INLINE pure #-}
-
-  WriterC f <*> WriterC a = WriterC $ \ w -> do
-    (w', f') <- f w
-    (w'', a') <- a w'
-    let fa = f' a'
-    fa `seq` pure (w'', fa)
-  {-# INLINE (<*>) #-}
-
-instance (Monad m, Monoid w) => Monad (WriterC w m) where
-  return = pure
-  {-# INLINE return #-}
-
-  m >>= f  = WriterC $ \w -> do
-    (w', a) <- runWriterC m w
-    runWriterC (f a) w'
-  {-# INLINE (>>=) #-}
-
-instance (Monoid w, Carrier sig m, Effect sig, Monad m) => Carrier (Writer w :+: sig) (WriterC w m) where
-  ret a = WriterC (\ w -> ret (w, a))
-  {-# INLINE ret #-}
-
-  eff op = WriterC (\ w -> handleSum (eff . handleState w runWriterC) (\case
-    Tell w'    k -> let w'' = mappend w w' in w'' `seq` runWriterC k w''
-    Listen   m k -> do
-      (w', a) <- runWriterC m mempty
-      let w'' = mappend w w'
-      w'' `seq` runWriterC (k w' a) w''
-    Censor f m k -> do
-      (w', a) <- runWriterC m mempty
-      let w'' = mappend w (f w')
-      w'' `seq` runWriterC (k a) w'')
-    op)
+newtype WriterC w m a = WriterC { runWriterC :: StateC w m a }
+  deriving (Alternative, Applicative, Functor, Monad, MonadFail, MonadIO, MonadPlus, MonadTrans)
+
+instance (Monoid w, Carrier sig m, Effect sig) => Carrier (Writer w :+: sig) (WriterC w m) where
+  eff (L (Tell w     k)) = WriterC $ do
+    modify (`mappend` w)
+    runWriterC k
+  eff (L (Listen   m k)) = WriterC $ do
+    w <- get
+    put (mempty :: w)
+    a <- runWriterC m
+    w' <- get
+    modify (mappend (w :: w))
+    runWriterC (k w' a)
+  eff (L (Censor f m k)) = WriterC $ do
+    w <- get
+    put (mempty :: w)
+    a <- runWriterC m
+    modify (mappend w . f)
+    runWriterC (k a)
```

Comment:
Defining the semantics of `WriterC` in terms of the semantics of `StateC` is lovely ✨